README
OpenGL-Array version 0.71_01
============================

This is a stand-alone distribution of OpenGL::Array
(which includes OpenGL::Matrix).  All of these were
implicitly loaded whenever the Perl OpenGL module was use-d.

This module should be compatible with OpenGL::Array
and OpenGL::Matrix from Perl OpenGL version 0.70 and can
be used as a replacement for that functionality.

The separate release is to enable backwards and forwards
compatibility between the legacy OpenGL module functionality
and the coming OpenGL::Modern module under development.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Devel::CheckOS      => 0
  ExtUtils::MakeMaker => 6.52
  Test::More          =>  0

Do not install this module with OpenGL.  It is only for
use separate from that module with OpenGL::Modern.

COPYRIGHT AND LICENCE

  Copyright (c) 1998,1999 Kenneth Albanowski. All rights reserved.
  Copyright (c) 2007 Bob Free. All rights reserved.
  Copyright (c) 2009-2013 Chris Marshall. All rights reserved.
  Copyright (C) 2017 Chris Marshall.  <chm@cpan.org>

  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.


Changes
Revision history for Perl extension OpenGL::Array.

0.71_01  TBD


0.71     2017-03-04 07:18:35-05:00
        - First official CPAN release


0.70_02  2017-03-03 09:40:46-05:00
        - Fix Apple includes for GL (uses <OpenGL/gl.h>)

0.70_01  2017-03-02 10:13:19-05:00
        - first stand alone version of OpenGL::Array
	- framework initialize with h2xs
	- code from Perl OpenGL 0.70


oga.pl
#!/usr/bin/perl -w
use strict;

use OpenGL qw(:all);

# Sample/Test app for OpenGL::Array functions
# by Bob "grafman" Free <grafman@graphcomp.com>
# http://graphcomp.com/opengl

use Time::HiRes qw( gettimeofday );


use constant PROGRAM_TITLE => "OpenGL::ARRAY Test App";

my @verts =
(
  -1.0, -1.3, -1.0,
  1.0, -1.3, -1.0,
  1.0, -1.3,  1.0,
  -1.0, -1.3,  1.0,

  -1.0,  1.3, -1.0,
  -1.0,  1.3,  1.0,
  1.0,  1.3,  1.0,
  1.0,  1.3, -1.0,

  -1.0, -1.0, -1.3,
  -1.0,  1.0, -1.3,
  1.0,  1.0, -1.3,
  1.0, -1.0, -1.3,

  1.3, -1.0, -1.0,
  1.3,  1.0, -1.0,
  1.3,  1.0,  1.0,
  1.3, -1.0,  1.0,

  -1.0, -1.0,  1.3,
  1.0, -1.0,  1.3,
  1.0,  1.0,  1.3,
  -1.0,  1.0,  1.3,

  -1.3, -1.0, -1.0,
  -1.3, -1.0,  1.0,
  -1.3,  1.0,  1.0,
  -1.3,  1.0, -1.0
);
@verts =
(
  0.0, 0.1, 0.2,
  1.0, 1.1, 1.2,
  2.0, 2.1, 2.2,
  3.0, 3.1, 3.2,

  10.0, 10.1, 10.2,
  11.0, 11.1, 11.2,
  12.0, 12.1, 12.2,
  13.0, 13.1, 13.2,

  20.0, 20.1, 20.2,
  21.0, 21.1, 21.2,
  22.0, 22.1, 22.2,
  13.0, 23.1, 23.2,

  30.0, 30.1, 30.2,
  31.0, 31.1, 31.2,
  32.0, 32.1, 32.2,
  33.0, 33.1, 33.2,

  40.0, 40.1, 40.2,
  41.0, 41.1, 41.2,
  42.0, 42.1, 42.2,
  43.0, 43.1, 43.2,

  50.0, 50.1, 50.2,
  51.0, 51.1, 51.2,
  52.0, 52.1, 52.2,
  53.0, 53.1, 53.2
);

my @xform =
(
  1.0, 0.0, 0.0, 1.0,
  0.0, 3.0, 0.0, 0.0,
  0.0, 0.0, 2.0, 1.0,
  0.0, 0.0, 0.0, 1.0
);
my $xform = OpenGL::Array->new_list(GL_FLOAT,@xform);

# Tests
my $loops = 2;
my $vertices = 4;
my $print_verts = $vertices < 5;

my($i,$start,$secs);
print "Initial data:\n";
print "Loops: $loops, Vertices: $vertices\n";
my $verts = init_verts($vertices);
print_verts($verts,$vertices) if ($print_verts);

$start = gettimeofday();
for ($i=0;$i<$loops;$i++)
{
  $verts->calc('1,+','3,*','2,*,1,+');
}
$secs = (gettimeofday()-$start)/$loops;
my $vps_calc = int($vertices/$secs);
print "Calc VPS: $vps_calc\n";
print_verts($verts,$vertices) if ($print_verts);


$verts = init_verts($vertices);
$start = gettimeofday();
for ($i=0;$i<$loops;$i++)
{
  $verts->affine($xform);
}
$secs = (gettimeofday()-$start)/$loops;
my $vps_gpu = int($vertices/$secs);
print "Affine VPS:  $vps_gpu\n";
print_verts($verts,$vertices) if ($print_verts);


sub init_verts
{
  my($count) = @_;

  my @verts = ();
  for (my $i=0;$i<$count;$i++)
  {
    push(@verts,"$i.0","$i.1","$i.2");
  }

  my $verts = OpenGL::Array->new_list(GL_FLOAT,@verts);
  die "new_list failed to return an OGA\n" if (!$verts);
  return $verts;
}

sub print_verts
{
  my($verts,$vertices) = @_;
  my $len = $vertices * 3;
  for(my $i=0;$i<$len;$i+=3)
  {
    my($x,$y,$z) = $verts->retrieve($i,3);
    printf("%8.3f,%8.3f,%8.3f\n",$x,$y,$z);
  }
  print "\n";
}


__END__


lib/OpenGL/Array.pm
package OpenGL::Array;

use 5.008000;
use strict;
use warnings;

require Exporter;
require DynaLoader;

our @ISA = qw(Exporter DynaLoader);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use OpenGL::Array ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.71_01';


bootstrap OpenGL::Array;

*OpenGL::Array::CLONE_SKIP  = sub { 1 };  # OpenGL::Array  is not thread safe
*OpenGL::Matrix::CLONE_SKIP = sub { 1 };  # OpenGL::Matrix is not thread safe

# Preloaded methods go here.

1;
__END__


Makefile.PL
use 5.008000;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use Devel::CheckOS qw(os_is);

my ( $DEFS, $INCS, $LIBS, $DYNS, $LDFROM, $EXES, $OPTS ); 

$EXES = [];
$DYNS = {};

if ( os_is( 'MSWin32' ) ) {
    $LIBS   = '-lopengl32 -lgdi32 -lmsimg32';
}
elsif ( os_is( 'Cygwin' ) ) {
    $LIBS = '-lGL -lX11';
}
elsif ( os_is( 'MacOSX' ) ) {
    $DYNS = { 'OTHERLDFLAGS' => '-framework OpenGL' };
}
else {
    $LIBS = '-lGL -lX11';
}

# This is the final build configuration
my $build_config =
{
    'NAME'              => 'OpenGL::Array',
    'VERSION_FROM'      => 'lib/OpenGL/Array.pm',  # finds $VERSION, requires EU::MM from perl >= 5.5
    'ABSTRACT_FROM'     => 'lib/OpenGL/Array.pod', # retrieve abstract from module
    'PREREQ_PM'         => {
        'Test::More' => 0
    },
    'CONFIGURE_REQUIRES' => {
        'Devel::CheckOS' => 0,
        'ExtUtils::MakeMaker' => 6.52,  # for CONFIGURE_REQUIRES
    },
    'LICENSE'           => 'perl',
    #'PM'                => { },
    'AUTHOR'            => "Chris Marshall ".'<chm at cpan dot org>',
    'OBJECT'            => 'opengl_array$(OBJ_EXT) gl_util$(OBJ_EXT) opengl_matrix$(OBJ_EXT)', 
    'XSPROTOARG'        => '-noprototypes',
    'DEFINE'            => $DEFS,
    'INC'               => $INCS,
    'LIBS'              => $LIBS,
    'dynamic_lib'       => $DYNS,
    'LDFROM'            => '$(OBJECT) '.(defined($LDFROM)?$LDFROM:''),
    'META_MERGE'        => {
        resources       => {
            homepage    => 'http://sourceforge.net/projects/pogl/',
            bugtracker  => 'http://sourceforge.net/tracker/?group_id=562483&atid=2281758',
            repository  => 'git://pogl.git.sourceforge.net/gitroot/pogl/pogl',
        },
    },
    'EXE_FILES'         => $EXES,
    'OPTIMIZE'          => $OPTS,
    #'clean'       => { FILES => }
};

WriteMakefile( %$build_config );                # Generate the Makefile
